components/automate-gateway/pkg/limiter/limiter.go

```diff
@@ -0,0 +1,90 @@
+package limiter
+
+import (
+	"errors"
+	"sync"
+
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+var ErrLimitExceeded = errors.New("Resource limit exceeded")
+
+type Limiter interface {
+	TakeTicket() (Ticket, error)
+}
+
+type Ticket interface {
+	Return()
+}
+
+type inflightRequestLimiter struct {
+	lock        sync.Mutex
+	maxRequests int
+	inflight    int
+	name        string
+}
+
+type inflightRequestLimiterTicket struct {
+	limiter  *inflightRequestLimiter
+	returned bool
+}
+
+func NewInflightRequestLimiter(name string, maxRequests int) Limiter {
+	return &inflightRequestLimiter{
+		name:        name,
+		maxRequests: maxRequests,
+	}
+}
+
+func (l *inflightRequestLimiter) TakeTicket() (Ticket, error) {
+	taken := false
+	cur := 0
+	max := 0
+	l.lock.Lock()
+	cur = l.inflight
+	max = l.maxRequests
+	if cur < max {
+		l.inflight++
+		taken = true
+	}
+	l.lock.Unlock()
+
+	if !taken {
+		err := status.Errorf(codes.ResourceExhausted,
+			"Resource limit exceeded for %s: cur=%d max=%d", l.name, cur, max)
```

Comment:
💭 Would cur and max ever have values that aren't equal?

Reply:
only if there are bugs

Reply:
I'd prefer to assume their presence. But feel free to ignore me if you feel differently.

Reply:
we don't need to log both, but we should log at least one as it makes debugging easier